CMakeLists.txt

```diff
@@ -1,283 +1,96 @@
-if(DEFINED RELEASE_CANDIDATE)
-	set(OBS_VERSION "${RELEASE_CANDIDATE}")
-	string(REPLACE "-rc" "." RC_SPLIT ${RELEASE_CANDIDATE})
-	string(REPLACE "." ";" RC_SPLIT ${RC_SPLIT})
-	message(WARNING "******************************************************************************\nRelease candidate deteced, OBS_VERSION is now: ${OBS_VERSION}\n******************************************************************************")
-	list(GET RC_SPLIT 0 OBS_RELEASE_CANDIDATE_MAJOR)
-	list(GET RC_SPLIT 1 OBS_RELEASE_CANDIDATE_MINOR)
-	list(GET RC_SPLIT 2 OBS_RELEASE_CANDIDATE_PATCH)
-	list(GET RC_SPLIT 3 OBS_RELEASE_CANDIDATE)
-else()
-	set(OBS_RELEASE_CANDIDATE_MAJOR 0)
-	set(OBS_RELEASE_CANDIDATE_MINOR 0)
-	set(OBS_RELEASE_CANDIDATE_PATCH 0)
-	set(OBS_RELEASE_CANDIDATE 0)
-endif()
-
-# Binary Versioning for Windows
-if(WIN32)
-	string(REPLACE "-" ";" UI_VERSION_SPLIT ${OBS_VERSION})
-	list(GET UI_VERSION_SPLIT 0 UI_VERSION)
-	string(REPLACE "." ";" UI_VERSION_SEMANTIC ${UI_VERSION})
-	list(GET UI_VERSION_SEMANTIC 0 UI_VERSION_MAJOR)
-	list(GET UI_VERSION_SEMANTIC 1 UI_VERSION_MINOR)
-	list(GET UI_VERSION_SEMANTIC 2 UI_VERSION_PATCH)
-	set(OBS_COMPANY_NAME "OBS")
-	set(OBS_PRODUCT_NAME "OBS Studio")
-	set(OBS_COMMENTS "Free and open source software for video recording and live streaming")
-	set(OBS_LEGAL_COPYRIGHT "(C) Hugh Bailey")
-	set(MODULE_DESCRIPTION "OBS Studio")
-	configure_file(UI/obs.rc.in ${PROJECT_BINARY_DIR}/obs.rc)
-endif()
-
-if(MSVC AND NOT EXISTS "${CMAKE_BINARY_DIR}/ALL_BUILD.vcxproj.user")
-	file(GENERATE
-		OUTPUT "${CMAKE_BINARY_DIR}/ALL_BUILD.vcxproj.user"
-		INPUT "${CMAKE_SOURCE_DIR}/cmake/ALL_BUILD.vcxproj.user.in")
+# Set default compiler flags
+include(CompilerConfig)
+
+# Prohibit in-source builds
+file(TO_CMAKE_PATH "${PROJECT_BINARY_DIR}/CMakeLists.txt" _LOC_PATH)
+if(EXISTS "${LOC_PATH}")
+  message(
+    FATAL_ERROR
+      "OBS: You cannot build in a source directory (or any directory with "
+      "CMakeLists.txt file). Please make a build subdirectory. Feel free to "
+      "remove CMakeCache.txt and CMakeFiles.")
 endif()
+unset(_LOC_PATH)
 
+# Allow selection of common build types via UI
 if(NOT CMAKE_BUILD_TYPE)
-        set(CMAKE_BUILD_TYPE RelWithDebInfo)
+  set(CMAKE_BUILD_TYPE
+      "RelWithDebInfo"
+      CACHE STRING
+            "OBS build type [Release, RelWithDebInfo, Debug, MinSizeRel]" FORCE)
+  set_property(CACHE CMAKE_BUILD_TYPE PROPERTY STRINGS Release RelWithDebInfo
+                                               Debug MinSizeRel)
 endif()
 
-set(CMAKE_CXX_STANDARD 17)
-set(CMAKE_CXX_STANDARD_REQUIRED YES)
-set(CMAKE_CXX_EXTENSIONS NO)
-
-if(${CMAKE_C_COMPILER_ID} MATCHES "Clang" OR ${CMAKE_CXX_COMPILER_ID} MATCHES "Clang")
-	set(CMAKE_COMPILER_IS_CLANG TRUE)
+# Global project options
+option(BUILD_FOR_DISTRIBUTION "Build for distribution (enables optimisations)"
+       OFF)
+option(ENABLE_UI "Enable building with UI (requires Qt)" ON)
+option(BUILD_QT5 "Enable building UI with Qt5" ON)
```

Comment:
I think `USE_QT5` would make more sense.

Though, if we try to support both Qt5 and Qt6 builds, perhaps this should be a constrained STRING instead?  With only two options, it's probably not much difference in work for either approach, so I don't mind too much which one we go with.  We'll just have to decide which one takes priority if we have booleans for both Qt5 and Qt6.

Reply:
FWIW I think once we're ready to move to Qt6, legacy Qt5 support shouldn't be something we stay concerned about.

Reply:
we will need to maintain Qt5 support for the duration of our ubuntu 20.04 support. And since it looks like debians qt maintainers stepped down, the likelihood of qt6 coming soon to ubuntu is low.

Reply:
Indeed, Linux platforms is the concern here.  We _can_ support both in CMake and code if needed.

Reply:
`ENABLE_UI` is the attribute who enables or not the build of the **UI** folder.
So naming it `USE_QT5` doesn't really make sense.

But a attribute named `USE_QT6` with an AUTO behavior like `ENABLE_RTMPS`, which:
- try to use Qt6 and fallback to Qt5 with AUTO
- use Qt6 with ON
- use Qt5 with OFF

Reply:
Sounds good - so let's rename it "USE_QT6" and have the "AUTO" behaviour be "OFF" until we have official support for it.

_However_ wouldn't this also mean that once we pull the switch that it **has** to break by default on Linux because of the whole "enforce all features by default" thing we discussed ITT ages ago? I'd be fine with that, but if that's not the case, we are introducing conflicting rules that I'd tried a lot to avoid again.

Reply:
Stupid me, I forgot to specify that `ENABLE_UI` and `USE_QT6` are two separated things.
The `ENABLE_UI` is ON but, the `USE_QT6` is AUTO by default so it automatically switch to Qt5 is Qt6 is not present.
And can be forced is set to ON or OFF.

Reply:
Sure, but what does "Qt6 not present" mean - Qt6 support in OBS' codebase, or Qt6 support by the OS?

Right now I'd have said "Qt6 support on OBS" and package maintainers would have to explicitly disable it (similar to how you'd need to explicitly disable Pipewire or Wayland now).

Reply:
If a Qt6 build provide features that a Qt5 build has not, yes Qt6 shall be enforced. And then the AUTO is not needed.
But if it's not case I don't know if enforcing to Qt6, is really necessary.

From what I read higher in the discussion, it seems that Qt6 availability on some distro is going to be messy.

If we default to Qt6 and AUTO is not needed, so it would be better to create `USE_QT5` which is OFF by default to provide a way to build OBS with Qt5.

Reply:
> "Qt6 not present" mean - Qt6 support in OBS' codebase, or Qt6 support by the OS?

OBS already mostly supports Qt6, and will fully support Qt6 with one more PR merged.  I took the flag to mean "Use Qt6 to build OBS" or "Use Qt5 to build OBS".  I didn't like "BUILD_QT5" because we aren't building Qt here.  You could use "BUILD_WITH_QT5", but "USE_QT5" is more succinct, to me.  I don't care too much, I just figured better to address it now rather than later.

Since it's an either/or an AUTO is probably fine, but I'll leave it up to @PatTheMav .

Reply:
Changed this to a constrained number variable as suggested by RytoEx.